File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
private:
    vector<vector<int>> result;
public:
    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        vector<int> current;
        backtrack(nums, current, target, 0);

        return result;
    }

    void backtrack(vector<int>& nums, vector<int>& current, int target, int index) {
        if(target == 0) {
            result.push_back(current);
            return;
        } else if (target < 0 || index == nums.size()) {
            return;
        }

        current.push_back(nums[index]);
        backtrack(nums, current, target - nums[index], index);
        current.pop_back();
        backtrack(nums, current, target, index + 1);
    }
};

File: Data Structures & Algorithms/two-integer-sum/submission-2.js
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} target
     * @return {number[]}
     */
    twoSum(nums, target) {
        let map = new Map()

        for(let i = 0; i < nums.length; i++) {
            let j = target - nums[i] // 7 - 4 = 3

            if(map.has(j)) { // 4
                return [map.get(j), i]
            }

            map.set(nums[i], i) // set map[4] = 0
        }

        return []
    }
}
